jetstreamext/getbatch.go

```diff
@@ -0,0 +1,299 @@
+		}
+		opts.Batch = batch
+		return nil
+	}
+}
+
+// GetLastMessagesFor fetches the last messages for the specified subjects from
+// the specified stream.
+// The function returns an iterator that can be used to iterate over the
+// messages.
+// It can be configured to fetch messages up to a certain stream sequence number or
+// time.
+func GetLastMessagesFor(ctx context.Context, js jetstream.JetStream, stream string, subjects []string, opts ...GetLastForOpt) (iter.Seq2[*jetstream.RawStreamMsg, error], error) {
+	if len(subjects) == 0 {
+		return nil, ErrSubjectRequired
+	}
+	reqOpts := &getLastBatchOpts{
+		MultiLastFor: subjects,
+	}
+
+	for _, opt := range opts {
+		if err := opt(reqOpts); err != nil {
+			return nil, err
+		}
+	}
+
+	reqJSON, err := json.Marshal(reqOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return getDirect(ctx, js, stream, reqJSON)
+}
+
+func getDirect(ctx context.Context, js jetstream.JetStream, stream string, req []byte) (iter.Seq2[*jetstream.RawStreamMsg, error], error) {
+	subj := getPrefixedSubject(js.Options(), fmt.Sprintf("DIRECT.GET.%s", stream))
+	eobSentinel := func(msg *nats.Msg) bool {
+		status, desc := msg.Header.Get(statusHdr), msg.Header.Get(descHdr)
+		return len(msg.Data) == 0 && status == statusNoContent && desc == descEOB
+	}
+	resp, err := natsext.RequestMany(ctx, js.Conn(), subj, req, natsext.RequestManySentinel(eobSentinel))
+	if err != nil {
+		return nil, err
+	}
+	return func(yield func(*jetstream.RawStreamMsg, error) bool) {
+		for msg, err := range resp {
+			if err != nil {
+				yield(nil, err)
+				return
+			}
+			rawMsg, err := convertDirectGetMsgResponseToMsg(msg)
+			if err != nil {
+				yield(nil, err)
+				return
+			}
+			if !yield(rawMsg, nil) {
+				break
+			}
+		}
+	}, nil
+}
+
+var (
+	statusHdr        = "Status"
+	descHdr          = "Description"
+	statusNoContent  = "204"
+	statusNoMessages = "404"
+	descEOB          = "EOB"
+)
+
+func convertDirectGetMsgResponseToMsg(msg *nats.Msg) (*jetstream.RawStreamMsg, error) {
+	if len(msg.Data) == 0 {
+		status := msg.Header.Get(statusHdr)
+		if status == statusNoMessages {
+			return nil, ErrNoMessages
+		}
+	}
+	if len(msg.Header) == 0 {
+		return nil, fmt.Errorf("%w: response should have headers", ErrInvalidResponse)
+	}
+	numPending := msg.Header.Get("Nats-Num-Pending")
+	if numPending == "" {
+		return nil, ErrBatchUnsupported
+	}
+	stream := msg.Header.Get(jetstream.StreamHeader)
+	if stream == "" {
+		return nil, fmt.Errorf("%w: missing stream header", ErrInvalidResponse)
+	}
+
+	seqStr := msg.Header.Get(jetstream.SequenceHeader)
+	if seqStr == "" {
+		return nil, fmt.Errorf("%w: missing sequence header", ErrInvalidResponse)
+	}
+	seq, err := strconv.ParseUint(seqStr, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("%w: invalid sequence header '%s': %v", ErrInvalidResponse, seqStr, err)
+	}
+	timeStr := msg.Header.Get(jetstream.TimeStampHeaer)
```

Comment:
typo. Should be `TimeStampHeader`

Reply:
It's a typo in `nats.go` unfortunately...